common/frame.go

```diff
@@ -220,50 +168,69 @@ func (f *Frame) clearLifecycle() {
 func (f *Frame) recalculateLifecycle() {
 	// Start with triggered events.
 	var events map[LifecycleEvent]bool = make(map[LifecycleEvent]bool)
-	f.lifecycleEventsMu.Lock()
-	for k, v := range f.lifecycleEvents {
-		events[k] = v
+	f.lifecycleEventsMu.RLock()
+	{
+		for k, v := range f.lifecycleEvents {
+			events[k] = v
+		}
 	}
-	f.lifecycleEventsMu.Unlock()
+	f.lifecycleEventsMu.RUnlock()
 
 	// Only consider a life cycle event as fired if it has triggered for all of subtree.
 	f.childFramesMu.RLock()
-	for child := range f.childFrames {
-		child.(*Frame).recalculateLifecycle()
-		for k := range events {
-			if !child.(*Frame).hasSubtreeLifecycleEventFired(k) {
-				delete(events, k)
+	{
```

Comment:
I think they make the protected area in the code clearer. You like indentation and don't use early returns pretty much, and I like this one. Can we agree on compromising some minor style differences in our code?

Reply:
> Nice work! This does indeed work much better with the Vistaprint script. It's a shame it's difficult to add a test for this, as we could really use some. Maybe some smaller unit tests for specific method changes could be added?

Thanks! I tried that as we discussed on Slack but it's challenging and I couldn't come up with good solutions while trying to debug thousands of logs—kinda mind stuck here.

> Although I'm not sure it fixes the `Cannot find context with specified id` issue, since it failed for me with that error:

Yes, we don't have a 100% fix and I don't know why. But I think we at least have 50%+ more improvement than before (in Tom's script). Also, this `Cannot find context with specified id` was for #49 but it looked like it solves some of the problems here as well, they're kind of connected. I can leave out that fix from this PR, though? And move it to #49.

> Also, please try to leave large logging changes for separate PRs, since they clutter the purpose of a "fix" PR. Refactorings are mostly fine if they're relevant to the fix, but otherwise those also belong in separate PRs. I mostly review commit by commit anyway, but it's difficult to see the final picture if the code is changed in the final diff.

Debugging is hard and I didn't want to throw away the work I've done. And doing it in another PR complicates things further for me.

I tried hard to make the commits easier for you to review by separating them into responsibilities, and I even added commits that fix the issue with explanations like: 

* [Fix cannot find context with specified id](https://github.com/grafana/xk6-browser/pull/129/commits/4f95b1be990f744c0bb4348ec3f2eb4e1ea4b860)
* [Fix cannot create frame session bug](https://github.com/grafana/xk6-browser/pull/129/commits/fab6b1aac529d14ddea0c4f203dfe97806234d0a)

🤷 😓 

Not like this [one](https://github.com/grafana/xk6-browser/pull/130/commits/4e2e427ac7860c9af42e26887eacf5a5a93eb79d) 😅

Reply:
Sure, if you feel strongly about this, of course we can compromise :smile:

To be fair, I can see how it would make the code clearer in cases where the lock is held between large blocks (which is actually a code smell since locks should be released as quickly as possible), as it highlights exactly where the lock is acquired and released. But doing it for <~10 LOCs or [one like in this case](https://github.com/grafana/xk6-browser/blob/235bc474b20634870b8ccf703b24505b31cd0762/common/page.go#L163-L167) just clutters the code and actually makes it _less_ readable. Especially since in most cases it can be refactored into a separate function that does `defer m.Unlock()`, or just done inline without the extra block.

My other issue with it is that it functionally changes behavior, as I need to be aware of an additional scope introduced for style purposes.

> You like indentation and don't use early returns pretty much

I think there was a misunderstanding :smile: I don't _like_ indentation, in fact I'm arguing against it right now... What I commented in your other PR was that refactoring 10 LOCs indented 4 levels deep just to remove 1 level of indentation doesn't make much sense, and just introduces noise. I actually quite like that kind of refactoring you've done in other places, including in this PR. I just suggested being pragmatic and not applying that for small blocks of code.

I don't have strong opinions on early returns, they're probably good in most situations, even though I prefer [single function exit points](https://wiki.c2.com/?SingleFunctionExitPoint) when possible (though this is impractical in Go with `err != nil` checks everywhere).

Anyway, it's fine, I won't bring it up again. I would just suggest being pragmatic about where it's used. I mean, if we want to apply it globally, it can become a linter, no?

Reply:
Good to know your thoughts, thanks for clarifying :)

> suggest being pragmatic about where it's used

+1, alright, I'll see what I can do afterward 😅 

> if we want to apply it globally, it can become a linter, no?

Some linters won't hurt :)